### DISCUSSION TITLE
Create audio effects options

### DISCUSSION THREAD

**Discussion by entonbiba** [katspaugh/wavesurfer.js#2688] [Ideas]
Allow for effects to be applied to the audio for example `fade in` `fade out` via options on render.

```
fadeInAudio: true,
fadeInLength: 10000,
fadeOutAudio: true,
fadeOutLength: 10000
```

**Comment by prakharpbuf** [comment]
I have been working on creating a DAW on browser using this library and React. 
I use the web audio API to apply these effects and wavesurfer.loadDecodedBuffer() to see the effective change in the waveform.

WaveSurfer allows us to add custom filters using wavesurfer.setFilters([]) but it does not re render the waveform to show the applied filter.

**Reply by prakharpbuf** [reply]
it will be nice to have an option for fadeIn/fadeOut effect on render. It will be nicer if the waveform also reflects the change otherwise I think it will worsen the UX imo

**Comment by entonbiba** [comment]
yes the effects would result in redrawing the peaks but the goal of these effects would be to make them work for both MediaElement and WebAudio

**Comment by katspaugh** [comment]
> it will be nice to have an option for fadeIn/fadeOut effect on render. It will be nicer if the waveform also reflects the change otherwise I think it will worsen the UX imo

It never occurred to me that effects may actually change the shape of the waveform, but it's a very valid point.
@entonbiba is right though, this won't work for MediaElement. It could probably work with MediaElementSourceNode if that could be piped into a buffer. Not sure.

Btw, what negative effects on the UX do you foresee, @prakharpbuf?

**Reply by prakharpbuf** [reply]
That was the only problem I faced when working on the DAW: the waveform did not reflect changes when adding effects.
I have a lot of effects right now in my test app like fading, normalization, noise reduction, plosive plugin, etc. But I'm not using the wavesurfer.setFilter() method anywhere in my app because the waveform does not reflect the changes.

instead, if I want to add a filter, I first apply the filter using an offline audio context then feed the resulting audio buffer to wavesurfer.

**Reply by prakharpbuf** [reply]
I'm guessing setFilter() adds the filter just before outputting the audio to the speakers? Like the waveform is drawn and everything then user filters are added then \<lastFilter\>.connect(ac.destination) ?

I think it might be useful if setFilter() adds the filters _before_ any script processors, gain nodes etc. and waveform is drawn from the resulting filters. Also every time someone adds/removes a filter drawBuffer() is called.

Edit: it's just a thought, I'm not entirely sure how it should be actually done

**Reply by katspaugh** [reply]
The waveform is drawn from an audio buffer that is returned by decodeAudioData. Then the same data (as a buffer node) is connected to the destination through a gain node.

So the waveform is rendered just one time when it fetches the audio.

I think your approach when you take audio data post-filtering and render another waveform from that makes sense.

**Comment by katspaugh** [comment]
Speaking of fade-in/fade-out, that should indeed already be possible with setFilters (basically an envelope on a gain node), but now I understand why #2689 was implemented via setVolume – to make it work for the MediaElement. Still think it should probably be a plugin, tho.

**Comment by katspaugh** [comment]
I've made an Envelope plugin, it's a graphical way to set fade-in and fade-out points, as well as to control the overall volume: https://wavesurfer-ts.pages.dev/tutorial/#/examples/envelope.js

**Reply by prakharpbuf** [reply]
wow! This looks really nice!

**Reply by prakharpbuf** [reply]
one question: Envelope plugin is using Web Audio gain node, so it won't be compatible with MediaElement right?

**Reply by katspaugh** [reply]
It’s built on top of the media element. The audio is passed to a web audio graph via a MediaElementSourceNode.
